### DISCUSSION TITLE
Can not pass data from snap to dapp with await

### DISCUSSION THREAD

**Discussion by pranavcpawar** [MetaMask/snaps#2685] [Q&A] [answered]
I'm working on a project where I'm creating a snap that generates a QR code string and then passes it to a dApp. To achieve this, I made a separate file called sdk.ts and imported it into index.tsx. However, I encountered the below error.
![image](https://github.com/user-attachments/assets/a577f6ce-8501-4de5-89a2-eeacb934bc63)
```
export const onRpcRequest: OnRpcRequestHandler = async ({
  origin,
  request,
}) => {
  switch (request.method) {
    case 'pairing': {
      await snap.request({
        method: 'snap_dialog',
        params: {
          type: 'confirmation',
          content: (
            <Box>
              <Text>Happy to have you on board! 🚀</Text>
            </Box>
          ),
        },
      });
      const qr = await sdk.initPairing();

      return { qrCode: qr };
    }
    default:
      throw new Error('Method not found.');
  }
};
```
Additionally, when I use this statement ``` const qr =  "hello world!" ``` instead of this ```const qr = await sdk.initPairing()``` it's working fine.

**Comment by ziad-saab** [comment]
@pranavcpawar can you provide more details on what `sdk.initPairing()` does? ideally can you share a link to a code repository that we can review?

**Reply by pranavcpawar** [reply]
Hello @ziad-saab, thanks for looking into this. The `sdk.initPairing` function basically pairs the snap account to a mobile app.
```
// sdk.ts

async function randomPairingId(): Promise<string> {
  const characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz';
  const charactersLength = characters.length;
  let result = '';
  
  for (let i = 0; i < 19; i++) {
    const randomIndex = Math.floor(Math.random() * charactersLength);
    result += characters[randomIndex];
  }
  
  return result;
}

export async function initPairing(): Promise<string> {
  try {
    const pairingId = await randomPairingId();

    // Generate encryption key pair using ECDH (Elliptic Curve Diffie-Hellman)
    const ecdh = createECDH('secp256k1');
    ecdh.generateKeys();
    const encPublicKey = ecdh.getPublicKey('hex');

    // Generate signing key pair
    const sign = createSign('SHA256');
    const signPrivateKey = randomBytes(32).toString('hex'); // Simulated private key
    sign.update(pairingId); // This is just a placeholder; adjust based on actual signing logic
    const signPublicKey = sign.sign(signPrivateKey, 'hex');

    const pairingCode = JSON.stringify({
      pairingId,
      webEncPublicKey: encPublicKey,
      signPublicKey,
    });

    return pairingCode;
  } catch (error: any) {
    throw new Error(error.message);
  }
}
```

**Reply by pranavcpawar** [reply]
hey @ziad-saab sir any updates?

**Reply by Mrtenz** [reply]
Hi @pranavcpawar, it seems like something in your `initPairing` function gets stuck or throws an error. Could you send a minimal reproduction so I can debug further?

**Reply by ziad-saab** [reply]
Hey @pranavcpawar,

To use the `createECDH` function in your code, or anything else from the `crypto` module, you have to [have polyfills enabled in your Snap configuration](https://docs.metamask.io/snaps/reference/cli/options/#polyfills).

If that fails or is already done, can you please provide a link to your complete source code so I can review further?

**Reply by pranavcpawar** [reply]
> Hi @pranavcpawar, it seems like something in your `initPairing` function gets stuck or throws an error. Could you send a minimal reproduction so I can debug further?

GM, @Mrtenz  yeah sure 
https://github.com/pranavcpawar/qr-snap

**Reply by pranavcpawar** [reply]
> Hey @pranavcpawar,
> 
> To use the `createECDH` function in your code, or anything else from the `crypto` module, you have to [have polyfills enabled in your Snap configuration](https://docs.metamask.io/snaps/reference/cli/options/#polyfills).
> 
> If that fails or is already done, can you please provide a link to your complete source code so I can review further?

GM @ziad-saab sir, I'm using [libsodium-wrappers](https://www.npmjs.com/package/libsodium-wrappers)
here's the codebase https://github.com/pranavcpawar/qr-snap

**Comment by ziad-saab** [comment]
Hey @pranavcpawar ,

I tried removing your `import _sodium from 'libsodium-wrappers';` and replacing it with a dynamic `const _sodium = (await import("libsodium-wrappers")).default` (in the functions), and I got it to work.

I don't know why it works this way and not with a regular import. I will get back to you on that, but meanwhile I hope this can unblock you.

**Comment by FrederikBolding** [comment, marked as answer]
Since `libsodium-wrappers` uses WebAssembly under the hood, you will need to request the `endowment:webassembly` permission as part of your `snap.manifest.json`: https://docs.metamask.io/snaps/reference/permissions/#endowmentwebassembly

After making that change, I was able to run a Snap with `libsodium-wrappers` imported.
